### ISSUE TITLE
No text-shaping for Tamil text

### ISSUE THREAD

**Issue by vizs** [koreader/koreader#9381]
* KOReader version: v2022.06
* Detected Device: taido_row
* OS: Android Marshmallow, api 23
* Build flavor: fdroid

#### Issue

KOReader does not shape Tamil text properly.  I tried setting the font to "Noto Sans" and "Noto Serif". 
 I will note that KOReader does seem to shape some text (eg. it shapes 0x0b95 0x0bcd properly) but overall, it is impossible to read the text since the text shaping is so poor.  See attached image comparing what the file should look like (from Emacs) and what it looks like in KOReader,

![out](https://user-images.githubusercontent.com/37809414/181434431-d8948355-6d0d-4d1d-8c09-b784a95bb524.png)

Desired - Left, KOReader - Right

Thanks for looking into this.

#### Steps to reproduce

1. Download the epub file from https://freetamilebooks.com/ebooks/some_months_in_england/.
2. Open it in KOReader and witness the broken text shaping.

(The epub file can be downloaded from the button that says "some_months_in_england.epub - Download" in the bottom.)

**Comment by Frenzie** [comment]
In the bottom menu, change font kerning from good to best.
![Screenshot_2022-07-28_08-51-31](https://user-images.githubusercontent.com/202757/181439937-b6041a03-5bad-4c74-93c7-b232698538a3.png)

**Comment by Frenzie** [comment]
A long-press on the option name will provide the explanation.

https://github.com/koreader/koreader/blob/2c952eca4d755f267e0e8ddea8a78753cedd0128/frontend/ui/data/creoptions.lua#L658-L665

**Comment by Frenzie** [comment]
@poire-z However, the book does specify `xml:lang="ta"`. Would it be possible to force best when that happens, or something along those lines?

**Comment by vizs** [comment]
Thanks but setting it to best did not make a difference.  See attached screenshot, 

![Screenshot-2022-07-28-13-44-24](https://user-images.githubusercontent.com/37809414/181456473-d1087848-bcfb-4aa7-936b-f55b5e186ed4.jpeg)

**Comment by Frenzie** [comment]
There's a great difference between good and best. Are you saying best is still incorrect?
<img src=https://user-images.githubusercontent.com/202757/181472659-d2150f0b-5f16-41b7-8d56-4f0eb3a76b4a.png width=45%> <img src=https://user-images.githubusercontent.com/202757/181472700-2b3f8798-3f8b-4870-8322-f7018ad04241.png width=45%>

**Comment by vizs** [comment]
Yes, exactly.  The red text in your screenshot should look like the attached,

![bad](https://user-images.githubusercontent.com/37809414/181482710-aab8e8d5-24d5-421d-95f8-c4bbdd787665.png)

This is just one of many blunders wrt text shaping in the document.  I am specifically stressing on text shaping because there are obvious blunders.  For example, consider the circled character in the image below,

![bad2](https://user-images.githubusercontent.com/37809414/181482711-969a97aa-a570-4002-9856-6a2b1ef02554.png)

It is composed of 0x09f 0x0bbf, if they are not shaped this is what it looks like: ட ி.  If it was shaped properly, it would be டி.

**Comment by Frenzie** [comment]
@poire-z Indeed it looks differently in Firefox. 
[ch028.xhtml.zip](https://github.com/koreader/koreader/files/9208175/ch028.xhtml.zip)

![image](https://user-images.githubusercontent.com/202757/181484214-4ad0afb4-b796-46d6-8b21-cd4edb7aeee5.png)

**Comment by poire-z** [comment]
> However, the book does specify xml:lang="ta". Would it be possible to force best when that happens, or something along those lines?

Not easily, the "kerning mode" is global and affects all instantiated fonts.
And somehow, that would be tedious as we would need to manage a list of language codes that would require harfbuzz.
Best to let that to the user. Our default is "best/harfbuzz", if the user toggles that to something else than "best", his mistake if he doesn't get "best" results :)

**Comment by poire-z** [comment]
> Indeed it looks differently in Firefox.

For shaping, there's not much in our hands :) we delegate all this to Harfbuzz.
Master still uses Harfbuzz 3.0.0 from one year ago.
You just bumped it to current 5.0.1, but it has not reached frontend (better keep that for after coming up v2022.07)
So, I'd like to see how it does with HB 5.0.1. Were you using that for your above tests ?

May be there were fixe over the last year for Tamil, but none of us but a Tamil reader would understand any of these issues.
https://github.com/harfbuzz/harfbuzz/issues?q=sort%3Aupdated-desc+tamil

What's in our hand and could introduce some bug is the text segmentation. It's best to use the Tamil font as the first/main/default font to not introduce segmentation as soon as a char is available in a top font.
Also, it may be bug in the font used for Tamil. Dunno if our chain of fallback fonts has some Tamil chars (we have Noto Sans Bengali and Devanagari, may be FreeSans or FreeSerif has some limited support for Tamic, dunno).
Or if you added a good Tamil font. (did you enable it in the fallback fonts for the UI ?)

**Comment by Frenzie** [comment]
> You just bumped it to current 5.0.1, but it has not reached frontend (better keep that for after coming up v2022.07)
So, I'd like to see how it does with HB 5.0.1. Were you using that for your above tests ?

I'm not 100% sure the screenshot I posted was taken with 5.0.1, but I did test both and they didn't differ in the top right character.

**Comment by poire-z** [comment]
OK, I looked at this EPUB.
In its CSS there is
```css
@font-face {
    src: url(../Fonts/Vijaya.ttf);
    panose-1: 2 11 6 4 2 2 2 2 2 4;
    font-family: "Vijaya"
    }
body { font-family: Vijaya; ...}
```
but that font is not included in the EPUB (dunno how it would look if it were).

So, the tamil codepoints are searched among our own fallback fonts.
And they can be found in FreeSans and FreeSerif - which are fonts with a large coverage, but probably not finely tuned for all the scripts, including Tamil. So, the result we see.
But that's not our bug, it's a FreeSans or FreeSerif bug - but they probably have a lot of such bugs (I think that's why we added Noto Sans Bengali).

But you can download [Noto Sans Tamil](https://fonts.google.com/noto/specimen/Noto+Sans+Tamil), and enable it as the main font or fallback font in the EPUB/crengine font menu - and as a fallback font for KOReader UI (file browser, ToC...) via `Gear>Device>Additional UI fallback fonts`. See #8722.
I think you would then get a more correct rendering:

![image](https://user-images.githubusercontent.com/24273478/181623810-bc6bfb24-48e0-4d31-8d7f-bab9cdf24e77.png)

**Comment by Frenzie** [comment]
Good catch!

**Comment by vizs** [comment]
Thanks, the latest screenshot you shared has perfect text shaping (and spacing).  BTW, I don't think there's a bug in harfbuzz wrt shaping Tamil text (Emacs uses it, and I only hit a bug when cairo was in the mix).

Do you have a reason why you don't allow fonts besides Noto Sans in the fallback section?  Wouldn't it be better if you say where to drop the font files in the additional UI fonts dialog, it took some jumping through hooks to get the info.  And the UI seems to pick the wrong weight of the font too: it first picked something lighter than Regular, so I deleted them all then it started to pick a bold font.

**Comment by poire-z** [comment]
> Do you have a reason why you don't allow fonts besides Noto Sans in the fallback section?

Because we want our UI to look nice, and using only fonts from the same Noto Sans family, we can assume they would all look nice to each others, so a filename or title like
`கணியம் 中国  עִבְרִי नागरी.epub` will look balanced.

>  Can you say where to drop the font files in the additional UI fonts dialog, it took some jumping through hoops to get the info.

This might be mentionned in the User Guide. The font path is different among devices. I assume a user will find the font/ subdir in its koreader/ directory.

> And the UI seems to pick the wrong weight of the font too: it first picked something lighter than Regular, so I deleted them all weights lighter than Regular, then it started to pick a bold font.

Yes, our UI font collector/classifier is a bit rudimentary. Not really worth making it more complex as we don't really need more that whan it does.
Except for this edge case where we would indeed pick and use the first thin/light/regular/medium we would meet... Well, I'd say let the user, seeing all other fonts have only regular and bold variations, pick only regular and bold variations of the fonts he is adding.

**Comment by vizs** [comment]
[ I sent the following as an e-mail reply but it didn't reach the tracker...  ]

>>Do you have a reason why you don't allow fonts besides Noto Sans in the fallback section?
>
>Because we want our UI to look nice, and using only fonts from the same Noto Sans family, we >can assume they would all look nice to each others, so a filename or title like
>கணியம் 中国 עִבְרִי नागरी.epub will look balanced.

I understand your concern but this is an artificial limitation: to get the additional fonts, it requires explicit action from the user so if the user wants to shoot their feet, let them do so.  [ I also prefer a serif face for Tamil but I can't have my cake.   ]

>>Can you say where to drop the font files in the additional UI fonts dialog, it took some jumping >>through hoops to get the info.
>
>This might be mentionned in the User Guide. The font path is different among devices. I assume >a user will find the font/ subdir in its koreader/ directory.

Unfortunately, I had to create the fonts/ directory myself here to get KOReader show Noto Tamil Sans here.  BTW, do you have plans to include Noto Sans Tamil in the default stack like you did for Devanagari and Bengali?  (You would need to do the same for other Indic languages too btw, I doubt FreeSans does text shaping.)

>>And the UI seems to pick the wrong weight of the font too: it first picked something lighter than >>Regular, so I deleted them all weights lighter than Regular, then it started to pick a bold font.>
>
>Yes, our UI font collector/classifier is a bit rudimentary. Not really worth making it more complex >as we don't really need more that whan it does.
>Except for this edge case where we would indeed pick and use the first >thin/light/regular/medium we would meet... Well, I'd say let the user, seeing all other fonts have >only regular and bold variations, pick only regular and bold variations of the fonts he is adding.

Hmm... would supporting a variable weight font worth it?  Noto Sans seems to come with that variant (at least, Noto Sans Tamil does).

**Comment by poire-z** [comment]
> I understand your concern but this is an artificial limitation: to get the additional fonts, it requires explicit action from the user so if the user wants to shoot their feet, let them do so. [ I also prefer a serif face for Tamil but I can't have my cake. ]

It's the same limitation that western users get: they can't set a latin serif font for their UI.
Our purpose is to get a nice and consistent enough UI. You don't spend your time in File manager or the Table of Contents. We care about fonts a bit more for the book rendering engine.
But see below.

>  BTW, do you have plans to include Noto Sans Tamil in the default stack like you did for Devanagari and Bengali? (You would need to do the same for other Indic languages too btw, I doubt FreeSans does text shaping.)

No, unless you and 3 others Tamil KOReader users show they exist, and cry loud enough :)
Although there are 75 miliions Tamil readers according to Wikipedia, so we can think about it.
But the more fonts we add, the larger the package is, and we don't want to be too fat.
Also, on Android devices, one usually gets all the Noto Sans font for free.
What would be nice is a additional fonts download subsystem, we talked about it somewhere, but it has not yet been implemented.

> Hmm... would supporting a variable weight font worth it? Noto Sans seems to come with that variant (at least, Noto Sans Tamil does).

It's another complex thing that nobody really needs, so I guess we won't spend time on this. It's just the UI after all.

Again, the aim is to have something nice, complete, and simple enough - and not have a whole bunch of new settings to tune UI fonts.
But, we have hooks to change fonts (don't really remember how that works, you may find some discussions among the issues), and since some weeks ago, support for user patches. So, somebody a bit interested and understanding Lua can live-patch KOReader to change a few things, including the list of fonts used I guess.

TLDR: we don't want to be Emacs :) But you can make _your_ KOReader be that with user patches.

**Comment by Frenzie** [comment]
@pazos provided a working poc plugin in a PR somewhere.

**Comment by pazos** [comment]
> But the more fonts we add, the larger the package is, and we don't want to be too fat.

I gave it a look a few months ago. There's nothing where we could squeeze a few hundred kb, except for translations.

On translations switching to binary form would save us about 300kb per language. Savings in compressed for, like in APK or deb files would be minimal.

> Also, on Android devices, one usually gets all the Noto Sans font for free.

They're behind a flag "enable system fonts", disabled by default,  because it is quite annoying to have 35 pages of fonts which contain mostly noto variants.

The same flag exist on other platform where system fonts are a thing.

**Comment by pazos** [comment]
> @pazos provided a working poc plugin in a PR somewhere.

Mmmm, nope unless you're talking about pulling fonts from google. In that case yep. There's a PR somewhete, just need to add a working api key, the rest should still still work

**Comment by vizs** [comment]
>It's the same limitation that western users get: they can't set a latin serif font for their UI.
Our purpose is to get a nice and consistent enough UI. You don't spend your time in File manager or the Table of Contents. We care about fonts a bit more for the book rendering engine.
But see below.

I am still not convinced FWIW.   Have you thought books (and fonts) that use PUA characters for scripts that are not in Unicode proper?   I'm not sure limiting for the sake of a consistent UI (which is highly appreciated, don't get me wrong) when the user knows what exactly they are stepping on sounds like a good idea.

> No, unless you and 3 others Tamil KOReader users show they exist, and cry loud enough :)
Although there are 75 miliions Tamil readers according to Wikipedia, so we can think about it.
But the more fonts we add, the larger the package is, and we don't want to be too fat.
Also, on Android devices, one usually gets all the Noto Sans font for free.
What would be nice is a additional fonts download subsystem, we talked about it somewhere, but it has not yet been implemented.

You prove my point, exactly.  The ship-with-the-fonts approach is not a scalable solution at all.  Perhaps, it would be better if we pointed out the "use the native fonts" option in the help page for additional UI font fallbacks?

>It's another complex thing that nobody really needs, so I guess we won't spend time on this. It's just the UI after all.

Fair enough.

>TLDR: we don't want to be Emacs :)

If only I could run Emacs as an actual app instead of caging it in a user interface that is clearly obsolete and not suited for my language but still being clinged onto for god knows what reason... (sorry for the mini rant.  if it wasn't clear, I was talking about ttys (and termux).)

>But you can make your KOReader be that with user patches.

I wish but Lua never fails me to give a headache when I touch it.  I would rather not go there.

Anyway, thanks for your help and patience.  If you have any other Tamil-related issues, feel free to ping me.  I would be more than happy to help.

**Comment by pazos** [comment]
> You prove my point, exactly. The ship-with-the-fonts approach is not a scalable solution at all.

It isn't intended to be scalable. It is intended to deliver a mvp on platforms where system fonts don't exist or are broken or obfuscated.
> Perhaps, it would be better if we pointed out the "use the native fonts" option in the help page for additional UI font fallbacks?

See above. native fonts only work on linux, mac, android and pocketbook. The rest of platforms rely on bundled fonts.

And we aim to deliver the same unicode coverage on all of them

**Comment by pazos** [comment]
also: download fonts from google is not a solution because it will require users to request their own api keys. 

But the plugin in question can be repurposed to pulm fonts from other servers. Maybe we want to self host, like with dicts?

**Comment by vizs** [comment]
>>Perhaps, it would be better if we pointed out the "use the native fonts" option in the help page for additional UI font fallbacks?

>See above. native fonts only work on linux, mac, android and pocketbook. The rest of platforms rely on bundled fonts.

>And we aim to deliver the same unicode coverage on all of them

But it wouldn't hurt to at least give a little hint towards the option right?

**Comment by poire-z** [comment]
Also, one pre-requisite to bundle a language specific font, expected from readers of the language, is to provide KOReader's translations for that language :)
But I hope we won't ship too many, because otherwise, the chain of fallback fonts would be quite long and would add some slowness (dunno how much) to reach the last ones like nerdfont.
(I have a MAX_FONT_NUM 16 in xtext.cpp, it looks like we could go up to 256, but I'm not sure if there's not a limitation somewhere.)
I think the `Device>Additional UI fallback fonts` is good enough for users to add ONE or TWO of the fonts they may need. We could just here have some download subsystem for Noto Sans fonts only.

**Comment by pazos** [comment]
> > > Perhaps, it would be better if we pointed out the "use the native fonts" option in the help page for additional UI font fallbacks?
> 
> > See above. native fonts only work on linux, mac, android and pocketbook. The rest of platforms rely on bundled fonts.
> 
> > And we aim to deliver the same unicode coverage on all of them
> 
> But it wouldn't hurt to at least give a little hint towards the option right?

Yeah. Please specify on which platforms is supported. Thanks!

**Comment by pazos** [comment]
> We could just here have some download subsystem for Noto Sans fonts only.

@poire-z: That sounds cool. We would need to host them to prevent the issue with api keys. Repurposing part of the previous PR is also possible. It will look like https://github.com/koreader/koreader/pull/6299/files#diff-9351a4f5ad0a88033fe88e570685f950139c23add1a4dcb33bc7876f26038168 but with other fonts :)

**Comment by vizs** [comment]
29 Jul 2022, 13:50 by ***@***.***:

>
>
>
>
>>
>> Do you have a reason why you don't allow fonts besides Noto Sans in the fallback section?
>>
>>
>
>
> Because we want our UI to look nice, and using only fonts from the same Noto Sans family, we can assume they would all look nice to each others, so a filename or title like
>  > கணியம் 中国  עִבְרִי नागरी.epub>  will look balanced.
>
>
>

I understand your concern but this is an artificial limitation: to get the additional fonts, it requires explicit action from the user so if the user wants to shoot their feet, let them do so.  [ I also prefer a serif face for Tamil but I can't have my cake.   ]


>>
>> Can you say where to drop the font files in the additional UI fonts dialog, it took some jumping through hoops to get the info.
>>
>>
>
> This might be mentionned in the User Guide. The font path is different among devices. I assume a user will find the font/ subdir in its koreader/ directory.
>
>

Unfortunately, I had to create the fonts/ directory myself here to get KOReader show Noto Tamil Sans here.  BTW, do you have plans to include Noto Sans Tamil in the default stack like you did for Devanagari and Bengali?  (You would need to do the same for other Indic languages too btw, I doubt FreeSans does text shaping.)

>>
>> And the UI seems to pick the wrong weight of the font too: it first picked something lighter than Regular, so I deleted them all weights lighter than Regular, then it started to pick a bold font.
>>
>>
>
>
> Yes, our UI font collector/classifier is a bit rudimentary. Not really worth making it more complex as we don't really need more that whan it does.
>  Except for this edge case where we would indeed pick and use the first thin/light/regular/medium we would meet... Well, I'd say let the user, seeing all other fonts have only regular and bold variations, pick only regular and bold variations of the fonts he is adding.
>
>
>

Hmm... would supporting a variable weight font worth it?  Then we can bake all the cake we need and eat it.
